src/OpenLoco/Paint/PaintMiscEntity.cpp

```diff
@@ -0,0 +1,208 @@
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level != 0)
+        {
+            return;
+        }
+
+        VehicleCrashParticle* particle = base->asVehicleCrashParticle();
+
+        loco_global<int32_t[5], 0x4FAAB4> vehicle_particle_base_sprites;
+        uint32_t imageId = vehicle_particle_base_sprites[particle->crashedSpriteBase] + particle->frame / 256;
+        imageId = Gfx::recolour2(imageId, particle->colourScheme);
+
+        session.addToPlotListAsParent(imageId, { 0, 0, particle->z }, { 1, 1, 0 });
+    }
+
+    static void paintExplosionCloudEntity(PaintSession& session, MiscBase* base)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level > 2)
+        {
+            return;
+        }
+        ExplosionCloud* particle = base->asExplosionCloud();
+        uint32_t imageId = 3372 + (particle->frame / 256);
+        session.addToPlotListAsParent(imageId, { 0, 0, particle->z }, { 1, 1, 0 });
+    }
+
+    static void paintSparkEntity(PaintSession& session, MiscBase* base)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level > 2)
+        {
+            return;
+        }
+        Spark* particle = base->asSpark();
+        uint32_t imageId = 3421 + (particle->frame / 256);
+        session.addToPlotListAsParent(imageId, { 0, 0, particle->z }, { 1, 1, 0 });
+    }
+
+    static void paintFireballEntity(PaintSession& session, MiscBase* base)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level > 2)
+        {
+            return;
+        }
+        Fireball* particle = base->asFireball();
+        uint32_t imageId = 3390 + (particle->frame / 256);
+        session.addToPlotListAsParent(imageId, { 0, 0, particle->z }, { 1, 1, 0 });
+    }
+
+    static void paintExplosionSmokeEntity(PaintSession& session, MiscBase* base)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level > 1)
+        {
+            return;
+        }
+        ExplosionSmoke* particle = base->asExplosionSmoke();
+        uint32_t imageId = 3362 + (particle->frame / 256);
+        session.addToPlotListAsParent(imageId, { 0, 0, particle->z }, { 1, 1, 0 });
+    }
+
+    static void paintSmokeEntity(PaintSession& session, MiscBase* base)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level > 1)
+        {
+            return;
+        }
+        Smoke* particle = base->as_smoke();
+        uint32_t imageId = 3465 + (particle->frame / 256);
+        session.addToPlotListAsParent(imageId, { 0, 0, particle->z }, { 1, 1, 0 });
+    }
+
+    // 0x00440325
+    void paintMiscEntity(PaintSession& session, MiscBase* base)
+    {
+        switch (base->getSubType())
+        {
+            case MiscThingType::exhaust: // 0
+            {
+                paintExhaustEntity(session, base);
```

Comment:
for all of these you can just pass in `base->as_exhaust()` or whatever. Hopefully the compiler gets clever enough to skip the double check of the type.

Reply:
Implemented in the following commit.